operator/src/main/java/oracle/kubernetes/operator/steps/InitializeWebhookIdentityStep.java

```diff
@@ -0,0 +1,187 @@
+  private static final File certFile = new File(WEBHOOK_DIR + "/config/webhookCert");
+  private static final File keyFile = new File(WEBHOOK_DIR + "/secrets/webhookKey");
+  public static final String WEBHOOK_KEY = "webhookKey";
+
+  /**
+   * Constructor for the InitializeWebhookIdentityStep.
+   * @param next Next step to be executed.
+   */
+  public InitializeWebhookIdentityStep(Step next) {
+    super(next);
+  }
+
+  @Override
+  public NextAction apply(Packet packet) {
+    try {
+      if (certFile.exists() && keyFile.exists()) {
+        // The webhook's ssl identity has already been created.
+        reuseIdentity();
+        return doNext(getNext(), packet);
+      } else {
+        // The webhook's ssl identity hasn't been created yet.
+        return createIdentity(packet);
+      }
+    } catch (Exception e) {
+      LOGGER.warning(WEBHOOK_IDENTITY_INITIALIZATION_FAILED, e.toString());
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void reuseIdentity() throws IOException {
+    // copy the certificate and key from the webhook's config map and secret
+    // to the locations the webhook runtime expects
+    FileUtils.copyFile(certFile, new File(WEBHOOK_CERTIFICATE));
+    FileUtils.copyFile(keyFile, new File(WEBHOOK_CERTIFICATE_KEY));
+  }
+
+  private NextAction createIdentity(Packet packet) throws Exception {
+    KeyPair keyPair = createKeyPair();
+    String key = convertToPEM(keyPair.getPrivate());
+    writeToFile(key, new File(WEBHOOK_CERTIFICATE_KEY));
+    X509Certificate cert = generateCertificate(WEBHOOK_CERTIFICATE, keyPair, SHA_256_WITH_RSA, COMMON_NAME,
+            CERTIFICATE_VALIDITY_DAYS);
+    writeToFile(getBase64Encoded(cert), new File(WEBHOOK_CERTIFICATE));
+    // put the new certificate in the webhook's config map so that it will be available
+    // the next time the webhook is started
+    return doNext(recordWebhookCert(WEBHOOK_CERTIFICATE, cert,
+            recordWebhookKey(key, getNext())), packet);
+  }
+
+  private static String convertToPEM(Object object) throws IOException {
+    StringWriter writer = new StringWriter();
+    try (JcaPEMWriter pemWriter = new JcaPEMWriter(writer)) {
+      pemWriter.writeObject(object);
+      pemWriter.flush();
+    }
+    return writer.toString();
+  }
+
+  private static void writeToFile(String content, File path) throws IOException {
+    path.getParentFile().mkdirs();
+    try (Writer wr = Files.newBufferedWriter(path.toPath())) {
+      wr.write(content);
+      wr.flush();
+    }
+  }
+
+  private static String getBase64Encoded(X509Certificate cert) throws IOException {
+    return Base64.getEncoder().encodeToString(convertToPEM(cert).getBytes());
+  }
+
+  private static Step recordWebhookCert(String certName, X509Certificate cert, Step next) throws IOException {
+    JsonPatchBuilder patchBuilder = Json.createPatchBuilder();
+    patchBuilder.add("/data/webhookCert", getBase64Encoded(cert));
```

Comment:
Thought... writing the cert to a CM and the key to a Secret just makes the concurrency problem harder... Why not just write both to the Secret? Then there will be one consistent update with both pieces of data.

Reply:
I was following the pattern that's similar to Operator's identity. In operator's case also, we write the cert to CM and key to a secret. Do you know why it's done that way in the operator code? Thanks.

Reply:
I believe that Lenny's scaling script reads the certificate from the CM.

Reply:
I changed to use the secret for both webhook cert and key.